### PR TITLE
Improve performance of TagMap in toString, expose helper API.

### PR DESCRIPTION
**Pull request by JakeWharton**
Calling `extensions(true)` and `get(Extension)` are expensive in terms of computation and allocation. This changes the `TagMap` to do an effectively-single-pass traversal of the underlying array appending to an existing `StringBuilder`. This method will also be used by generated code, and the implementation of `toString` in the message adapter has been updated to reflect more closely the behavior of what generated code would look like.

Without a method like this, generating `toString` in code is very verbose.

### CODE REVIEW

File: wire-runtime/src/test/java/com/squareup/wire/WireTest.java [hunk 1]
```diff
@@ -211,7 +211,7 @@ public void testExtensionEnum() throws Exception {
     // Original value shows up as an extension.
     assertThat(msg.toString()).contains("squareup.protos.simple.nested_enum_ext=BAZ");
     // New value is unknown in the tag map.
-    assertThat(newMsg.toString()).contains("ExternalMessage{129=[17]}");
+    assertThat(newMsg.toString()).contains("ExternalMessage{129=17}");
```
Comment (danrice-square) [inline wire-runtime/src/test/java/com/squareup/wire/WireTest.java]: Why this change?

File: wire-runtime/src/main/java/com/squareup/wire/RuntimeMessageAdapter.java [hunk 2]
```diff
@@ -165,31 +165,22 @@ B newBuilder(M value) {
...
-      if (seenValue) sb.append(", ");
-      if (extension.isUnknown()) {
-        sb.append(extension.getTag());
-      } else {
-        sb.append(extension.getName());
+      if (value != null) {
+        sb.append(", ")
+            .append(fieldBinding.name)
+            .append('=')
+            .append(fieldBinding.redacted ? "██" : value);
```
Comment (danrice-square) [inline wire-runtime/src/main/java/com/squareup/wire/RuntimeMessageAdapter.java]: Define a constant for the redacted string with a comment as to what char is being used.

File: wire-runtime/src/test/java/com/squareup/wire/WireTest.java (hunk 1 above)
Comment (JakeWharton) [inline wire-runtime/src/test/java/com/squareup/wire/WireTest.java]: The extension is not a repeated one. It represents a single enum value.

File: wire-runtime/src/main/java/com/squareup/wire/RuntimeMessageAdapter.java [hunk 3]
```diff
@@ -165,31 +165,22 @@ B newBuilder(M value) {
...
       }
-      sb.append('=').append(message.tagMap.get(extension));
-      seenValue = true;
     }
-    sb.append('}');
-    return sb.toString();
+    message.tagMap.appendToString(sb);
+    sb.ensureCapacity(name.length() + 2); // Make sure next two chars fit, for empty objects.
+    sb.setCharAt(name.length() - 1, name.charAt(name.length() - 1));
```
Comment (danrice-square) [inline wire-runtime/src/main/java/com/squareup/wire/RuntimeMessageAdapter.java]: -1, handling of initial comma is too tricky.

File: wire-runtime/src/main/java/com/squareup/wire/RuntimeMessageAdapter.java [hunk 4]
```diff
@@ -166,30 +168,20 @@ B newBuilder(M value) {
...
       }
-      sb.append('=').append(message.tagMap.get(extension));
-      seenValue = true;
     }
-    sb.append('}');
-    return sb.toString();
+    message.tagMap.appendToString(sb);
+
+    // Replace leading comma with class name and opening brace.
+    sb.replace(0, 2, messageType.getSimpleName() + '{');
```
Comment (swankjesse) [inline wire-runtime/src/main/java/com/squareup/wire/RuntimeMessageAdapter.java]: what if the message is empty?

File: wire-runtime/src/main/java/com/squareup/wire/TagMap.java [hunk 5]
```diff
@@ -227,6 +227,31 @@ static void transcode(List<Object> list, Extension<?, ?> sourceExtension,
...
+          .append(extension.isUnknown() ? extension.getTag() : extension.getName())
+          .append("=");
+      int end = runEnd(i);
+      if (end > i + 2) {
+        builder.append('[');
+        for (int start = i; i < end; i += 2) {
+          if (i > start) builder.append(", ");
+          builder.append(array[i + 1]);
+        }
+        i -= 2; // Undo double-increment.
```
Comment (swankjesse) [inline wire-runtime/src/main/java/com/squareup/wire/TagMap.java]: instead of this, move the `i += 2` above to line 250?

File: wire-runtime/src/main/java/com/squareup/wire/RuntimeMessageAdapter.java (hunk 4 above)
Comment (JakeWharton) [inline wire-runtime/src/main/java/com/squareup/wire/RuntimeMessageAdapter.java]: `replace` will ensure there is enough capacity. I can write a test that shows it.

File: wire-runtime/src/main/java/com/squareup/wire/TagMap.java (hunk 5 above)
Comment (swankjesse) [inline wire-runtime/src/main/java/com/squareup/wire/TagMap.java]: ("double increment" is confusing because it's incrementing by 2)

**Comment by swankjesse** [comment]
LGTM
